### ISSUE TITLE
Why am I no longer getting generated synthetic features?

### ISSUE THREAD

**Issue by CBrauer** [EpistasisLab/tpot#766]
Hey,

Back in August 2018, I ran TPOT and got really good results using version 0.9.3.  At that time, I was surprised to see that TPOT synthesized new features.  My dataset contains five features, and TPOT generated an additional 19 new features, all of which were more important than my original features.

I just re-ran the same script, using TPOT version 0.9.5, and no new features were added.  How can I get TPOT to generate new features like before?

The extrapolated pipeline that TPOT generated today is:

```
import numpy as np
import pandas as pd
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler

# NOTE: Make sure that the class is labeled 'target' in the data file
tpot_data = pd.read_csv('PATH/TO/DATA/FILE', sep='COLUMN_SEPARATOR', dtype=np.float64)
features = tpot_data.drop('target', axis=1).values
training_features, testing_features, training_target, testing_target = \
            train_test_split(features, tpot_data['target'].values, random_state=None)

# Average CV score on the training set was:0.9104662944552648
exported_pipeline = make_pipeline(
    StandardScaler(),
    RandomForestClassifier(bootstrap=False, criterion="gini", max_features=0.55, min_samples_leaf=1, min_samples_split=2, n_estimators=100)
)

exported_pipeline.fit(training_features, training_target)
results = exported_pipeline.predict(testing_features)
```
The extrapolated pipeline that was generated last August, using TPOT version 0.9.3 was:

```
# Score on the training set was:0.9377722249192809
exported_pipeline = make_pipeline(
    make_union(
        FunctionTransformer(copy),
        make_union(
            Nystroem(gamma=0.8500000000000001,
                     kernel="cosine",
                     n_components=6),
            make_union(
                make_pipeline(
                    StackingEstimator(estimator=ExtraTreesClassifier(bootstrap=False,
                                                                     criterion="entropy",
                                                                     max_features=0.3,
                                                                     min_samples_leaf=8,
                                                                     min_samples_split=11,
                                                                     n_estimators=100)),
                    Normalizer(norm="l1")
                ),
                FunctionTransformer(copy)
            )
        )
    ),
    ExtraTreesClassifier(bootstrap=False,
                         criterion="gini",
                         max_features=0.7000000000000001,
                         min_samples_leaf=1,
                         min_samples_split=2,
                         n_estimators=500)
)

```
Keep up the great work.  I love TPOT.

Charles

**Comment by weixuanfu** [comment]
From v0.9.3 to v0.9.5, TPOT added/refined some operators in the default configuration although TPOT still has `StackingEstimator` for generating synthetic features. But the result may not be reproduced for 2 reasons: 

1. the default configuration is updated so that the initial population (of pipeline) should be changed
2. the `random_state` is `None` in the scripts, so the old result cannot be reproduced via a different random seed.

You can try to [customize the operators's configuration](https://epistasislab.github.io/tpot/using/#customizing-tpots-operators-and-parameters) to limit the search space of TPOT. In this case, you can only keep `ExtraTreesClassifier`, `Normalizer` and `Nystroem` from default configuration. But the fitted pipeline maybe the pipeline you had before or a new one with better accuracy.

**Comment by CBrauer** [comment]
Hi Weixuan,

Thanks for the reply.  Is there a way to tell TPOT to use StackingEstimator? The new features generated better results for me.

Charles

**Comment by weixuanfu** [comment]
Hmm, the current version of default TPOT can only evaluate the pipeline that are randomly generated in genetics programming. But increasing population size or offspring size may generate more pipelines with `StackingEstimator`.

**Comment by CBrauer** [comment]
I just completed a run with generations = population = 100.  No feature synthesis!  It looks like you guys are turning your back on feature synthesis, which is a axiom of autoML.

You didn't answer my question.  How can I tell TPOT to use StackingEstimator?  I cloned TPOT from GitHUb and I looked at ./tpot/tpot/config/classifier.py.  I assume this is the default configuration.  If so, why is there no StackingEstimator in this file?

**Comment by weixuanfu** [comment]
Sorry for misunderstanding. We still have feature synthesis and it can be used if one step in a pipeline is an estimator (classifier/regressor) and is not the last step then we must add its guess as synthetic feature(s). And `StackingEstimator` is not in the default configuration but TPOT can use it no matter what includes in default configuration. (see [here](https://github.com/EpistasisLab/tpot/blob/master/tpot/export_utils.py#L320-L328)). You can check the `evaluated_individuals_` attribute of fitted `TPOTClassifier` or `TPOTRegressor`, some evaluated pipeline should include `StackingEstimator`. Below is a demo to check if `StackingEstimator` was used in evaluated pipelines. I tested this demo based on TPOT v0.9.5 

```python
from tpot import TPOTClassifier
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
import numpy as np
from tpot.export_utils import generate_pipeline_code, expr_to_tree
from deap import creator

iris = load_iris()
X_train, X_test, y_train, y_test = train_test_split(iris.data.astype(np.float64),
    iris.target.astype(np.float64), train_size=0.75, test_size=0.25, random_state=42)

tpot = TPOTClassifier(generations=5, population_size=50, verbosity=2,random_state=42)
tpot.fit(X_train, y_train)


for pipeline_string in sorted(tpot.evaluated_individuals_.keys()):
    deap_pipeline = creator.Individual.from_string(pipeline_string, tpot._pset)
    sklearn_pipeline_str = generate_pipeline_code(expr_to_tree(deap_pipeline, tpot._pset), tpot.operators)
    if sklearn_pipeline_str.count('StackingEstimator'):
        print(sklearn_pipeline_str)
        print('evaluated_pipeline_scores: {}\n'.format(tpot.evaluated_individuals_[pipeline_string]))
```

In the Pareto fronts of TPOT, it assume that less-complex pipelines (with less operator numbers in a pipeline) will generalize better. So `StackingEstimator` may increase pipeline complexity but the overall performance (average CV scores) is not too competitive with some less complex pipeline, which may be the reason that TPOT didn't select it as best pipeline in the end.

**Comment by rhiever** [comment]
@CBrauer, one thing to clarify here is that TPOT uses a stochastic search algorithm. Thus, TPOT may discover the pipeline you discovered previously in one run and find an entirely different pipeline in another run. (Assuming there are multiple pipelines that can achieve similar CV performance.)

If the pipeline with the stacking estimator does indeed perform better than the simpler pipelines, then TPOT will discover it given enough run-time. Hence @weixuanfu's suggestions.

**Comment by CBrauer** [comment]
Thank you guys. I appreciate your responses.

Another way of putting it: “Hey Charlie, be thankful we didn’t need to do feature synthesis in your case”.
Keep up the good work, I love TPOT.

Charles

**Comment by rhiever** [comment]
It's possible... :-)

To your earlier question, we explored the engineering challenge of allowing users to *force* TPOT to explore pipelines following a certain pipeline format. In that case, you could (for example) force TPOT to use feature synthesis. That unfortunately proved to be a larger engineering effort than our team could handle, but it's still on our wish list if the programming time/funding ever became available for it.
